src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace Symfony\Component\AutoMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\AutoMapper\Exception\RuntimeException;
+use Symfony\Component\AutoMapper\Attributes\MapIf;
+use Symfony\Component\AutoMapper\Attributes\MapTo;
+use Symfony\Component\AutoMapper\Attributes\MapWith;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Object to object mapper.
+ *
+ * @experimental
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class AutoMapper implements AutoMapperInterface
+{
+    public function __construct(private ?PropertyAccessorInterface $propertyAccessor = null, private ?ContainerInterface $serviceLocator = null)
+    {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function map(object $object, object|string|null $to = null): mixed
+    {
+        $refl = new \ReflectionClass($object);
+
+        if (!$to) {
+            $to = $this->getAttribute($refl, MapTo::class, true)->to;
+        }
+
+        $arguments = [];
+        if (is_object($to)) {
+            $toRefl = new \ReflectionClass(get_class($to));
+            $mapped = $to;
+        } else {
+            $toRefl = new \ReflectionClass($to);
+            $mapped = $toRefl->newInstanceWithoutConstructor();
+        }
+
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $arguments[$parameter->getName()] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
+            $propertyName = $property->getName();
+            $mapTo = $this->getAttribute($property, MapTo::class)?->to ?? $propertyName;
```

Comment:
It's an annoying edge-case, but if people need to be able to map to different objects in different situations... e.g. `Product` maps to `ProductApiDto` sometimes and `ProductFormDto` other times...

Reply:
We have two solutions for that:

1. forbid that and make it mandatory to use another DTO
2. add a `group` property to the attributes as we already have for the Validator and the Serializer component (probably the best option).

Reply:
I have another idea which is to compose with attributes:

```
#[MapIf(fn(mixed $value, object $object): bool => true, mapTo: new MapTo(to: Bar::class))]
#[MapIf(fn(mixed $value, object $object): bool => false, mapTo: new MapTo(to: Baz::class))]
class Product {}
```

Thoughts?

Reply:
I kinda like that... 

Possible to have `MapTo` still as the top-level attribute? And put the conditional inside of it? I'm just thinking it looks nice to always have `MapTo` above the class... and then occasionally, it's conditional

```php
#[MapTo(to: Bar::class, if: new MapIf(fn()....)]
```

Reply:
Indeed this is even better thanks!

Reply:
no they're not valid. It's "pseudo-code" actually you would use static functions or a service name.

Reply:
Yes it's pseudo code, i assume later it would be able to handle expression using the ExpressionLanguage component